internal/services/mobilenetwork/mobile_network_attached_data_network_data_source.go

```diff
@@ -0,0 +1,215 @@
+package mobilenetwork
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/location"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/mobilenetwork/2022-11-01/attacheddatanetwork"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/mobilenetwork/2022-11-01/packetcoredataplane"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type AttachedDataNetworkDataSource struct{}
+
+var _ sdk.DataSource = AttachedDataNetworkDataSource{}
+
+func (r AttachedDataNetworkDataSource) ResourceType() string {
+	return "azurerm_mobile_network_attached_data_network"
+}
+
+func (r AttachedDataNetworkDataSource) ModelObject() interface{} {
+	return &AttachedDataNetworkModel{}
+}
+
+func (r AttachedDataNetworkDataSource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return attacheddatanetwork.ValidateAttachedDataNetworkID
+}
+
+func (r AttachedDataNetworkDataSource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"mobile_network_data_network_name": {
```

Comment:
Shall we rename this to `name`, or do we want to rename the corresponding field in the model from `Name` to `MobileNetworkDataNetworkName` to be consistent?

Reply:
On reflection, given that this needs to be the same value as the corresponding Mobile Network Data Network's name value, we should in fact use the ID for that here and extract the name from it. This ensures the value is consistent, provides additional validation, and the value can be pulled from a reference allowing an explicit dependency between the resources.
```suggestion
		"mobile_network_data_network_id": {
```

Reply:
The reason to use name instead of id is the id format of `azurerm_mobile_network_attached_data_network` does not contain `mobileNetworkName`, the response/model does not contain it either...

attached_data_network_id:`/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MobileNetwork/packetCoreControlPlanes/{packetCoreControlPlaneName}/packetCoreDataPlanes/{packetCoreDataPlaneName}/attachedDataNetworks/{attachedDataNetworkName}`

data_network_id: `/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.MobileNetwork/mobileNetworks/{mobileNetworkName}/dataNetworks/{dataNetworkName}`

Reply:
<img width="782" alt="Screenshot 2023-06-29 at 11 15 13" src="https://github.com/hashicorp/terraform-provider-azurerm/assets/51212351/13177663-83e7-4f67-801b-930b3bc3fa72">

Here is the dependency between these resources, black arrow means dependency, yellow arrow means id path. As there is a `N:1` relation between `azurerm_mobile_network_site` and `azurerm_mobile_network_packet_core_control_plane`, itt seems there is no way get the id of the id of `azurerm_mobile_network_data_network`. WDYT?